### ISSUE TITLE
Blank Node IDs in REST API

### ISSUE THREAD

**Issue by iddan** [eclipse/rdf4j#1395]
I'm using the RDF4J REST API for managing an RDF store containing blank nodes. When I add a blank node with the `ADD` action they are assigned a different ID than the one assigned to it in the payload. This ID is not communicated back so I have no way to perform a `GET` on this node.
I would expect either the `ADD` action will allow preserving blank node identifiers or will communicate back the generated IDs.

**Comment by jeenbroekstra** [comment]
This is by design. Blank nodes are just that: blank. The node id only has meaning in its local context (for example, in a file, if the same blank node is referenced in two places), but is not intended to reference or identify that node _outside_ that particular context. Rdf4j replaces node identifiers when storing blank nodes to avoid possible clashes: e.g. if you were to upload two documents that use the same node identifier, and rdf4j _didn't_ reassign ids, it would lead the two nodes being unified (which would be incorrect).

Communicating the blank node id that rdf4j uses internally back to you is pointless: it's only an internal reference for the store itself, you can't use it for anything else (for example, you can't use it in a SPARQL query to query that particular node). 

The TL;DR is that if you have a use case where you need some way to actually reference or identify nodes, you should not use blank nodes, but create IRIs instead.

**Comment by iddan** [comment]
The RIO parser exposes an option to preserve blank node identifiers http://docs.rdf4j.org/programming/#_example_blank_node_preservation
If it's possible in Java I don't see a reason it shouldn't be in REST

**Comment by jeenbroekstra** [comment]
Well yes, but I figured I'd start by explaining why, typically, preserving blank node ids is a Bad Idea(tm). 

To some extent, what you want is already possible. You can configure your rdf4j server to preserve blank nodes by means of a system property. Start up with `-Dorg.eclipse.rdf4j.rio.preserve_bnode_ids=true` and it will not replace blank nodes when you upload rdf data. Use at your peril.

For more fine-grained control of parser settings over REST (e.g. controlling this per request than only by means of a startup switch), there is an open issue (#446) for an improvement to communicate client-side parser configuration to the server. However, it's not been given a lot of attention sofar, that issue.
